src/syncback/property_filter.rs

```diff
@@ -0,0 +1,112 @@
+use std::collections::HashMap;
+
+use rbx_dom_weak::{types::Variant, Instance};
+use rbx_reflection::{PropertyKind, PropertySerialization, Scriptability};
+
+use crate::{variant_eq::variant_eq, Project};
+
+/// Returns a map of properties from `inst` that are both allowed under the
+/// user-provided settings, are not their default value, and serialize.
+pub fn filter_properties<'inst>(
+    project: &Project,
+    inst: &'inst Instance,
+) -> HashMap<&'inst str, &'inst Variant> {
+    let mut map: Vec<(&str, &Variant)> = Vec::with_capacity(inst.properties.len());
+    filter_properties_preallocated(project, inst, &mut map);
+
+    map.into_iter().collect()
+}
+
+/// Fills `allocation` with a list of properties from `inst` that are
+/// user-provided settings, are not their default value, and serialize.
+pub fn filter_properties_preallocated<'inst>(
+    project: &Project,
+    inst: &'inst Instance,
+    allocation: &mut Vec<(&'inst str, &'inst Variant)>,
+) {
+    let sync_unscriptable = project
+        .syncback_rules
+        .as_ref()
+        .and_then(|s| s.sync_unscriptable)
+        .unwrap_or_default();
+
+    let class_data = rbx_reflection_database::get()
+        .classes
+        .get(inst.class.as_str());
+
+    let predicate = |prop_name: &String, prop_value: &Variant| {
+        // We don't want to serialize Ref or UniqueId properties in JSON files
+        if matches!(prop_value, Variant::Ref(_) | Variant::UniqueId(_)) {
+            return true;
+        }
+        if !should_property_serialize(&inst.class, prop_name) {
+            return true;
+        }
+        if !sync_unscriptable {
+            if let Some(data) = class_data {
+                if let Some(prop_data) = data.properties.get(prop_name.as_str()) {
+                    if matches!(prop_data.scriptability, Scriptability::None) {
+                        return true;
+                    }
+                }
+            }
+        }
+        false
+    };
+
+    if let Some(class_data) = class_data {
+        let defaults = &class_data.default_properties;
+        for (name, value) in &inst.properties {
+            if predicate(name, value) {
+                continue;
+            }
+            if let Some(default) = defaults.get(name.as_str()) {
+                if !variant_eq(value, default) {
+                    allocation.push((name, value));
+                }
```

Comment:
If I set a property to a non-default value and sync it back, then set the property back to its default value and sync it back, the property will be filtered out and the final result will be the first, non-default value. I'm not sure at this point how to fix this

We should generally be very careful in this function, and ensure that we don't erroneously filter out any properties!

Reply:
So to fix this, meta, JSON model, and project syncback middlewares need to *remove* properties if they exist in the old property map and if they have default values. 

I'm not sure what is the best way to go about this, but one idea I have for a first step is to make the property filter fn aware of both the old and new instance, and compare old and new values when filtering default properties (i.e. only filter a default property if both old and new values are at the default, or when there is no old instance).

Alternatively, the middlewares themselves could be responsible for filtering properties, instead of the property filter fn. Is it important for hashing that default properties are filtered out?

Reply:
If memory serves, the reason the default properties are filtered out for hashing is to guard against hashes being different if a model file is updated (i.e. manually resaved from Studio) but not changed otherwise. The idea is that we only care about properties that aren't their default values when determining if an Instance is different.

That said, I think it's probably fair to have JSON middlewares handle this themselves, but I'm not sure what the best way to approach it is. I'm worried about things getting more spaghetti than they are, so I'm hesitant to add more functions lol

Reply:
I've put together a tentative fix for this at https://github.com/UpliftGames/rojo/pull/6